src/SqlPersistence/Outbox/OutboxPersister.cs

```diff
@@ -33,7 +33,26 @@ public Task<OutboxTransaction> BeginTransaction(ContextBag context)
         var transaction = outboxTransactionFactory();
         transaction.Prepare(context);
         // we always need to avoid using async/await in here so that the transaction scope can float!
-        return transaction.Begin(context);
+        return BeginTransactionInternal(transaction, context);
+    }
+
+    private static async Task<OutboxTransaction> BeginTransactionInternal(ISqlOutboxTransaction transaction, ContextBag context)
+    {
+        OutboxTransaction result = null;
+        try
+        {
+            result = await transaction.Begin(context).ConfigureAwait(false);
+            return result;
+        }
+        catch (Exception e)
+        {
+            // A method that returns something that is disposable should not throw during the creation
+            // of the disposable resource. If it does the compiler generated code will not dispose anything
+            // therefore we need to dispose here to prevent the connection being returned to the pool being
+            // in a zombie state.
+            result?.Dispose();
```

Comment:
Shouldn't that be `transaction` and not `result?`? If we throw in `Begin` the result will be null and we won't dispose of the resources.